[IMP] website_hr_recruitment: design improvement

karlosvalverde opened the pull request:
Old `website_hr_recruitment` templates looked outdated, so they have
been improved in terms of usability and design, by applying the following
changes:

- Old images used in `/jobs` and `/jobs/detail/` have been replaced by
new ones, taken from the Odoo official profile in social media.

- Old left column in `/jobs` has been removed and the filters
contained in this column have been moved into a topbar.

- Filters in the topbar now work as dropdown items, optimised for small
viewports, and are disabled in views != '/jobs'.

- A searchbar has been added to the previous topbar.

- A `Connect with us` section has been added to the right column of
`/jobs`, including a separator, links and social media icons.

- Job entries in `/jobs` now show the city of the office, its country 
and its department, as shown in the filters.

- In `/jobs/*`, old header with a breadcrumb has been replaced by
the topbar seen above, including a link to previous page in the 
navigation flow.

- Design in `/jobs/detail/` has been improved, including a redesign of
the `Our Product` block or the image gallery at the end of the page.

- Header in `jobs/apply/*` has been improved and includes a column with
the address of the job position.

- In `jobs/apply/*`, the design of the form group has been revamped and 
also has been added a right column including a snippet with the contact 
information of the recruiter tracking the job application process.

- Overall design has been improved in the `/job-thank-you` view, 
including the recruiter snippet seen above.

- Job description is now editable, not truncated and also shown in the
main section of the Job Detail view.

- A pager has been added, showing 12 entries by page.

- `Remote` filterable location has been included, and can be set either
by using the backend or directly with the website builder.

- Every new job entry has now a default short description, editable.

- If there is either no available job or not any valid job after the job
search, the `No jobs available` view has also been improved.

- Illustrations with dynamic colours have been included in `/jobs` (`No
jobs available`), `/jobs/detail` (`Our Product` block) and 
`/job-thank-you`.

- Syntax has been adapted to BS5.

Co-authored with @bso-odoo 

Task 2406626

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/67913).

karlosvalverde commented:
Hi @stefanorigano 

You can have a look to this PR. :wink:

stefanorigano commented on addons/website_hr_recruitment/static/src/scss/website_hr_recruitment.scss:
```diff
@@ -1,24 +1,89 @@
+            }
+        }
+        .card {
+            padding: 15px;
```
Better to use default bootstrap classes (eg. `card-body`).
If it's not possible at last use the default grid metric (16px)

stefanorigano commented on addons/website_hr_recruitment/static/src/scss/website_hr_recruitment.scss:
```diff
@@ -1,24 +1,89 @@
+        .card {
+            padding: 15px;
+            box-shadow: 0 1px 1px 0 rgba(0, 0, 0, 0.17);
+            @include o_jobs_has_shadows();
```
If I'm not wrong these 2 mixins are used here and for the following selector only.
If confirmed you can avoid create a mixins and just do:
```scss
.card, .o_jobs_not_card {
 mixin stuff
}
.o_jobs_not_card {
 other stuff
}
```

stefanorigano commented on addons/website_hr_recruitment/static/src/scss/website_hr_recruitment.scss:
```diff
@@ -37,3 +102,48 @@
+        border: $border-width solid gray('400');
+        @include o-bg-color(gray('white'), $with-extras: false);
+        @include border-radius($dropdown-border-radius);
+        @include hover-focus {
```
Is all this really needed ? Can't we use the default design instead?

Moreover this approach:
- generate inconsistencies across modules
- overrides the theme design (even worst, may lead to a mix of both designs)
- overrides variable customisations written by the user using the HTML editor
- ...

I'm aware that currently other modules customize these component, but we should standardise rather than jeopardise even more.

stefanorigano commented on addons/website_hr_recruitment/static/src/scss/website_hr_recruitment.scss:
```diff
@@ -37,3 +102,48 @@
+    background: linear-gradient(150deg, #875a7b 20%, #62495b 80%) !important;
+}
+
+.s_image_gallery {
```
This will affect **all** `s_image_gallery` snippets.

stefanorigano commented on addons/website_hr_recruitment/static/src/scss/website_hr_recruitment.scss:
```diff
@@ -37,3 +102,48 @@
+    }
+}
+
+.o_hr_index_banner {
```
What's this?

stefanorigano commented on addons/website_hr_recruitment/static/src/scss/website_hr_recruitment.scss:
```diff
@@ -1,24 +1,89 @@
+    }
+}
+@mixin o_jobs_has_color {
+    &.o_jobs_has_color {
```
Not sure about this design :/
A card that gets a primary coloured border on hover only... it's not be consistent with all the other modules that use either cards or lists.

karlosvalverde commented:
Before having the latest changes in the task validated, I'll be temporarily setting this PR as a draft. 😉

Guillaume-gdi commented on addons/website_hr_recruitment/views/website_hr_recruitment_templates.xml:
```diff
@@ -7,92 +7,320 @@
+                                    <li><i class="fa fa-phone fa-fw mr-2"></i><span class="o_force_ltr"><a href="tel:+1 (650) 691-3277">+1 (650) 691-3277</a></span></li>
+                                </ul>
+                                <!-- Social media -->
+                                <div class="s_share text-left" data-name="Social Media">
+                                    <h5 class="s_share_title d-none">Follow us</h5>
+                                    <a href="/website/social/facebook" class="s_share_facebook" target="_blank">
+                                        <i class="fa fa-facebook btn btn-outline-primary btn-lg shadow-sm" style="border-radius: 100px;"></i>
+                                    </a>
+                                    <a href="/website/social/twitter" class="s_share_twitter" target="_blank">
+                                        <i class="fa fa-twitter btn btn-outline-primary btn-lg shadow-sm" style="border-radius: 100px;"></i>
+                                    </a>
+                                    <a href="/website/social/linkedin" class="s_share_linkedin" target="_blank">
+                                        <i class="fa fa-linkedin btn btn-outline-primary btn-lg shadow-sm" style="border-radius: 100px;"></i>
+                                    </a>
+                                </div>
```
I would replace with this:
```suggestion
                                <div class="s_social_media text-left" data-snippet="s_social_media" data-name="Social Media" contenteditable="false">
                                <h5 class="s_social_media_title d-none" contenteditable="true">Follow us</h5>
                                <a href="/website/social/facebook" class="s_social_media_facebook" target="_blank">
                                    <i class="fa fa-facebook rounded-circle shadow-sm" contenteditable="false">​</i>
                                </a>
                                <a href="/website/social/twitter" class="s_social_media_twitter" target="_blank">
                                    <i class="fa fa-twitter rounded-circle shadow-sm" contenteditable="false">​</i>
                                </a>
                                <a href="/website/social/linkedin" class="s_social_media_linkedin" target="_blank" contenteditable="true">
                                    <i class="fa fa-linkedin rounded-circle shadow-sm d-none" contenteditable="false">​</i>
                                </a>
                                <a href="/website/social/instagram" class="s_social_media_instagram d-none" target="_blank">
                                    <i class="fa fa-instagram rounded-circle shadow-sm" contenteditable="false">​</i>
                                </a>
                                <a href="/website/social/youtube" class="s_social_media_youtube d-none" target="_blank">
                                    <i class="fa fa-youtube rounded-circle shadow-sm" contenteditable="false">​</i>
                                </a>
                                <a href="/website/social/github" class="s_social_media_github d-none" target="_blank">
                                    <i class="fa fa-github rounded-circle shadow-sm" contenteditable="false">​</i>
                                </a>
                                <a href="/" class="text-800 d-none">
                                    <i class="fa fa-home rounded-circle shadow-sm" contenteditable="false">​</i>
                                </a>
                            </div>
```

Using the new `social_media` block, the user will be able to manage links to these social networks more easily. The `s_share` block should only be used for sharing on social networks.

karlosvalverde replied:
Indeed. I'm replacing it. Thanks!

karlosvalverde commented:
@stefanorigano Could you have a look into this PR?

stefanorigano commented on addons/website_hr_recruitment/static/src/scss/website_hr_recruitment.scss:
```diff
@@ -1,24 +1,29 @@
+                box-shadow: 0 6px 10px 0 rgba(0, 0, 0, 0.1), 0 2px 2px 0 rgba(0, 0, 0, 0.05);
+            }
+        }
+        .o_jobs_unpublished {
```
Missing empty line before the selector... could you check the other lines too?

karlosvalverde replied:
Done!

stefanorigano reviewed (changes requested):
Hello, some remarks.

karlosvalverde commented on addons/website_hr_recruitment/static/src/scss/website_hr_recruitment.scss:
```diff
@@ -1,24 +1,29 @@
-                color: theme-color('primary');
-                margin-right: 8px;
+                color: theme-color('secondary');
+                margin-right: .5rem;
```
In this case, this change should be kept, especially since we are using a widget in the 't-field' as a 't-option', which invokes a fa icon by default, but it doesn't allow to add any class customisation, unless the widget is being modified to do so.

https://github.com/odoo-dev/odoo/blob/32238cfd27ab23379a795a26b2b1f8770085034f/addons/website_hr_recruitment/views/website_hr_recruitment_templates.xml#L258

karlosvalverde commented on addons/website_hr_recruitment/static/src/scss/website_hr_recruitment.scss:
```diff
@@ -1,24 +1,29 @@
             }
         }
+        .card {
+            padding: $card-spacer-x;
```
Done!

karlosvalverde commented on addons/website_hr_recruitment/views/website_hr_recruitment_templates.xml:
```diff
@@ -7,92 +7,317 @@
+                                <hr class="w-100 mx-auto" style="border-top-width: 1px; border-top-style: solid; border-top-color: rgb(206, 212, 218) !important;"/>
+                            </div>
+                            <p id="connect">
+                                <ul class="list-unstyled text-nowrap">
```
Indeed. I just replaced it with a '<div>' tag.

karlosvalverde commented on addons/website_hr_recruitment/static/src/scss/website_hr_recruitment.scss:
```diff
@@ -1,24 +1,29 @@
+    }
+    #jobs_grid_right {
+        ul.flex-column > li.nav-item > a {
+            padding: .5rem 1rem;
```
Indeed. It is not needed anymore. I'm deleting it. 👍

karlosvalverde commented on addons/website_hr_recruitment/views/website_hr_recruitment_templates.xml:
```diff
@@ -7,92 +7,317 @@
                                     </div>
                                 </div>
+                                <div class="col-lg-12 text-center pt24">
+                                    <h5 class="font-weight-light text-reset">There are currently no open job opportunities,</h5>
```
It looks like `font-weight-light` does not override user customisations on the text... 🤔 

In any case, I'm deleting `text-reset` and assigning a `<font class="text-black">` instead.

stefanorigano replied:
> In any case, I'm deleting text-reset and assigning a <font class="text-black"> instead.

again, with `text-black` too,  _if o_cc2 bg is dark the title will not be readable_

karlosvalverde commented:
@stefanorigano About the "still valid" comments above:

- In the first case, unless the user changes the icon colour to a white bg, by default the icon takes the primary value and the bg remains white, no matter what colour is in the bg. However, I am unable to see any readability issues related to a change of the navbar design.

- In the second issue, the ´.o_hr_index_banner´ class was defined to a gradient bg used in a banner, which is no longer used here (FP told us to remove banners).

What do you think?

stefanorigano commented:
> In the first case, unless the user changes the icon colour to a white bg, by default the icon takes the primary value and the bg remains white, no matter what colour is in the bg. However, I am unable to see any readability issues related to a change of the navbar design.

I think that I have already said what I think  [here](https://github.com/odoo/odoo/pull/67913/files#r745661779) 😕
Once again,  these customisations override the default design making themes/variable customisations ineffective.
You forced the border-color to `gray('400')` , the button's border-radius to `$dropdown-border-radius`, and the border and text colors in :hover state too.

I have the feeling that you design on the default theme/palette only, without considering that the user can actually change colors.
If I want to create a "dark theme" and use solid black buttons, **I will be stuck with your white ones instead**.
Moreover, since my theme is dark, I would probably set $primary as a very light-color, and **this color will not be readable on your white background** (the readability issue). Detail; `hover-focus` is discontinued. 

> In the second issue, the ´.o_hr_index_banner´ class was defined to a gradient bg used in a banner, which is no longer used here (FP told us to remove banners).

My comment [here](https://github.com/odoo/odoo/pull/67913/files#r745679497) was about the implementation.
Once again, you're currently targeting **ALL** `s_image_gallery` across the website. Either you scope the selector within jobs **OR** you  apply the changes in the snippet's scss file.

karlosvalverde commented:
> > In the first case, unless the user changes the icon colour to a white bg, by default the icon takes the primary value and the bg remains white, no matter what colour is in the bg. However, I am unable to see any readability issues related to a change of the navbar design.
> 
> I think that I have already said what I think [here](https://github.com/odoo/odoo/pull/67913/files#r745661779) 😕 Once again, these customisations override the default design making themes/variable customisations ineffective. You forced the border-color to `gray('400')` , the button's border-radius to `$dropdown-border-radius`, and the border and text colors in :hover state too.
> 
> I have the feeling that you design on the default theme/palette only, without considering that the user can actually change colors. If I want to create a "dark theme" and use solid black buttons, **I will be stuck with your white ones instead**. Moreover, since my theme is dark, I would probably set $primary as a very light-color, and **this color will not be readable on your white background** (the readability issue). Detail; `hover-focus` is discontinued.
> 
> > In the second issue, the ´.o_hr_index_banner´ class was defined to a gradient bg used in a banner, which is no longer used here (FP told us to remove banners).
> 
> My comment [here](https://github.com/odoo/odoo/pull/67913/files#r745679497) was about the implementation. Once again, you're currently targeting **ALL** `s_image_gallery` across the website. Either you scope the selector within jobs **OR** you apply the changes in the snippet's scss file.

After some consideration, I've decided to drop the previous scss related to the conflictive dropdown elements and to use `.btn-outline-primary` classes instead, which is properly supported by our colour system, so it improves readability and usability.

About the aforementioned `.text-black` class in the `no-jobs` view, checking our colour Combination guide, I've finally replaced it with a `.text-muted` class, which allows headings to be adapted in relation to the parent bg colour.

Finally, `s_image_gallery` customisations have now been wrapped inside a specific `.o_jobs_image_gallery` class.

Thanks a lot for your review! 👍

karlosvalverde commented:
Hi @bso-odoo 

Could you have a look into this PR? I think we could start a technical review 😉

bso-odoo commented on addons/website_hr_recruitment/views/website_hr_recruitment_templates.xml:
```diff
@@ -448,6 +371,185 @@
+                            <span class="w-100 o_force_ltr d-block">Remote</span>
+                        </div>
+                    </t>
+                    <span t-field="job.address_id" t-options='{"widget": "contact", "class": "fw-light", "fields": ["address"],"no_tag_br": False}'/>
```
Missing space before `"no_tag_br"`.

karlosvalverde replied:
Done 👍

bso-odoo commented on addons/website_hr_recruitment/views/website_hr_recruitment_templates.xml:
```diff
@@ -448,6 +371,185 @@
+            <!-- Title -->
+            <div t-if="job" class="row pt24">
+                <div class="col-lg-8">
+                    <h1 class="text-left">
+                        <span t-field="job.name"/>
+                    </h1>
+                    <h5 class="fw-light text-left mt-2 text-muted">
+                        Job Application Form
+                    </h5>
```
With BS5, aren't we supposed to use start & end instead of left & right ? (also happens in other blocks in this file)

karlosvalverde replied:
Good one! I just fixed it, thanks! 👍

bso-odoo commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -2485,7 +2485,7 @@ const Many2oneUserValueWidget = SelectUserValueWidget.extend({
         options.domain = JSON.parse(options.domain);
         options.nullText = $target[0].dataset['nullText'] ||
-            JSON.parse($target[0].dataset['oeContactOptions'])['null_text'];
+            JSON.parse($target[0].dataset['oeContactOptions'] || '{}')['null_text'];
```
It would be nicer to move each of the fix of this commit to their origin commit - and ultimately remove this commit.

karlosvalverde replied:
Do you mean to split the commit in as many commits as fixes it includes?

bso-odoo replied:
Yes, then put each one inside one of the previous commits.

karlosvalverde replied:
Done!

Maybe you would like to add a more detailed description to the `fix handling of possibly missing json` commit. 😉

bso-odoo replied:
No need: it should just be "fixup" within the initial commit where that line appeared.

karlosvalverde replied:
Alright, I'm squashing it 👍

bso-odoo commented on addons/website_hr_recruitment/data/hr_job_demo.xml:
```diff
@@ -5,19 +5,23 @@
         <field name="website_description" type="html">
             <!-- Description text and ratings -->
-            <section class="pt32">
+            <section class="pt0">
```
Commit title: `[IMP] website_hr_recruitment: improve pages design` ?

karlosvalverde replied:
Done 👍

bso-odoo commented on addons/website_hr_recruitment/controllers/main.py:
```diff
@@ -3,45 +3,55 @@
 
 from odoo import http, _
 from odoo.addons.http_routing.models.ir_http import slug
+from odoo.osv.expression import AND
```
Commit title: `[IMP] website_hr_recruitment: add pager and search` ?

karlosvalverde replied:
Done 👍

bso-odoo commented on addons/website_hr_recruitment/controllers/main.py:
```diff
@@ -5,6 +5,7 @@
 from odoo.addons.http_routing.models.ir_http import slug
 from odoo.osv.expression import AND
 from odoo.http import request
+from odoo.tools.misc import groupby
```
Commit title: `[IMP] website_hr_recruitment: add filters` ?
Also, it would be nicer if the template updates related to the filters were introduced inside this commit.

karlosvalverde replied:
Indeed, done 👍

bso-odoo commented on addons/website/tests/test_fuzzy.py:
```diff
@@ -225,7 +225,7 @@ def test_06_case_insensitive_results(self):
     def test_07_no_fuzzy_for_mostly_number(self):
         """ Ensures exact match is used when search contains mostly numbers. """
-        self._create_page('Product P7935432254U7 page', 'Product P7935432254U7', '/numberpage')
+        self._create_page('Product P7935432254U7 page', 'Product P7935432254U7 kangaroo shoes', '/numberpage')
```
Commit message:
```
[FIX] website: adapt fuzzy test

Changes done in this task introduce new data that involve the word
"product" the fuzzy test is based on.

This commit adapts the test so that its results are based on words
that are less likely to appear in default data.

task-2406626
```
?

karlosvalverde replied:
It looks more descriptive and easy to read the changes on the commit 👍 

Done 👌

bso-odoo reviewed (commented):
Here are a few comments.

bso-odoo commented:
Extra comment: did you make sure the images can be distributed ?

karlosvalverde commented on addons/web_editor/models/ir_qweb_fields.py:
```diff
@@ -216,6 +220,11 @@ def from_html(self, model, field, element):
         many2one_id = int(element.get('data-oe-many2one-id'))
+        allow_reset = element.get('data-oe-many2one-allowreset')
+        if allow_reset and many2one_id == 0:
+            # reset the id of the many2one
```
Done 👍

karlosvalverde commented:
> Extra comment: did you make sure the images can be distributed ?

Actually, those images are taken from our social media channels, so I think they can be distributed (otherwise, we could not be able to share them on social media 😉).

About the illustrations, they come from our `iap-apps` illustration library.

karlosvalverde commented on addons/website_hr_recruitment/data/hr_job_demo.xml:
```diff
@@ -199,28 +206,25 @@
             </section>
             <!-- Photos -->
-            <section class="pt24 pb16">
+            <section class="o_jobs_image_gallery s_image_gallery pt24 pb24 o_fake_not_editable o_colored_level o_grid o_spc-medium" data-vcss="001" data-columns="3" style="overflow: hidden;" data-snippet="s_images_wall" data-name="Images Wall">
```
Indeed, that class was not necessary, the same as `o_colored_level`.

I also realised that the photo gallery was not updated on the description page of a new job instance, which is now fixed. 👍

karlosvalverde commented:
@bso-odoo I think everything looks better now. 😉 

Should we move forward with the task?

bso-odoo commented:
> @bso-odoo I think everything looks better now. wink
> 
> Should we move forward with the task?

Maybe one more little change before that (sorry ^_^)
It feels like `[FIX] website_hr_recruitment: tour selector fix` and `[FIX] website_hr_recruitment: default_website_description loading fix` should be squashed into their respective commits.

For `[FIX] website_hr_recruitment: job.description fix` I am not sure: if the problem existed before this PR, we should keep it distinct (but then we need to write a [proper message](https://www.odoo.com/documentation/15.0/contributing/development/coding_guidelines.html#commit-message-header) => e.g. `[FIX] website_hr_recruitment: avoid adding empty lines to job description`) as it will possibly need to be moved to a stable version. Otherwise (if the problem it solves it related to changes done inside this PR), it should be squashed into the commit that introduced the problem.

karlosvalverde commented:
> > @bso-odoo I think everything looks better now. wink
> > Should we move forward with the task?
> 
> Maybe one more little change before that (sorry ^_^) It feels like `[FIX] website_hr_recruitment: tour selector fix` and `[FIX] website_hr_recruitment: default_website_description loading fix` should be squashed into their respective commits.
> 
> For `[FIX] website_hr_recruitment: job.description fix` I am not sure: if the problem existed before this PR, we should keep it distinct (but then we need to write a [proper message](https://www.odoo.com/documentation/15.0/contributing/development/coding_guidelines.html#commit-message-header) => e.g. `[FIX] website_hr_recruitment: avoid adding empty lines to job description`) as it will possibly need to be moved to a stable version. Otherwise (if the problem it solves it related to changes done inside this PR), it should be squashed into the commit that introduced the problem.

I've squashed `[FIX] website_hr_recruitment: tour selector fix` and `[FIX] website_hr_recruitment: default_website_description loading fix` to `[IMP] website_hr_recruitment: improve pages design`, since it is the commit whose changes are related to, and I have dropped `[FIX] website_hr_recruitment: job.description fix`, because I notice that the problem has been solved in standard.


Thanks for the review! 👍

JKE-be commented on addons/website_hr_recruitment/controllers/main.py:
```diff
@@ -3,74 +3,190 @@
 
     @http.route([
         '/jobs',
-        '/jobs/country/<model("res.country"):country>',
-        '/jobs/department/<model("hr.department"):department>',
-        '/jobs/country/<model("res.country"):country>/department/<model("hr.department"):department>',
-        '/jobs/office/<int:office_id>',
-        '/jobs/country/<model("res.country"):country>/office/<int:office_id>',
-        '/jobs/department/<model("hr.department"):department>/office/<int:office_id>',
-        '/jobs/country/<model("res.country"):country>/department/<model("hr.department"):department>/office/<int:office_id>',
-        '/jobs/employment_type/<int:contract_type_id>',
-        '/jobs/country/<model("res.country"):country>/employment_type/<int:contract_type_id>',
-        '/jobs/department/<model("hr.department"):department>/employment_type/<int:contract_type_id>',
-        '/jobs/office/<int:office_id>/employment_type/<int:contract_type_id>',
-        '/jobs/country/<model("res.country"):country>/department/<model("hr.department"):department>/employment_type/<int:contract_type_id>',
-        '/jobs/country/<model("res.country"):country>/office/<int:office_id>/employment_type/<int:contract_type_id>',
-        '/jobs/department/<model("hr.department"):department>/office/<int:office_id>/employment_type/<int:contract_type_id>',
-        '/jobs/country/<model("res.country"):country>/department/<model("hr.department"):department>/office/<int:office_id>/employment_type/<int:contract_type_id>',
```
quid about retrocompatibility for old links?
Some very specific can probably just redirect to more global page.
But we don't want a 404 for each these rules e.g. /jobs/country/Belgium-20

Users work for seo, it is not to be broken after each upgrade.
![image](https://user-images.githubusercontent.com/4848797/191279610-adfc71ee-f569-45e3-8762-91c0311606de.png)

karlosvalverde replied:
Indeed. We have adapted the routes in order to maintain backward compatibility.

The changes are included in the commit below:

https://github.com/odoo/odoo/pull/67913/commits/bdd855924ba4faef07f48b4b4c0fe6790faa90ee

Thanks!

@bso-odoo

bso-odoo commented:
@karlosvalverde @qsm-odoo Rebased on 16.0: `hr.job` has been extracted into its own python file.
I however fear this is too late now for 16.0 since we add one field to the model. :cry:

bso-odoo commented:
@qsm-odoo Updated the commit messages and included the last request from GMF to force the order of the filter dropdowns on screen.
I asked him to confirm he is ok with the current order of filters - matching the one of the options. The answer is pending. At worst, the change will be renumbering the `priority` attribute of the templates.

bso-odoo commented:
> @bso-odoo I can already see the co-authored lines are still not right grimacing

@qsm-odoo Added the missing colons...

qsm-odoo commented on addons/website_hr_recruitment/views/website_hr_recruitment_templates.xml:
```diff
@@ -3,81 +3,126 @@
-                                </p>
-                            </div>
-                        </div>
+        <div id="wrap" class="o_website_hr_recruitment_jobs_list o_cc o_cc2">
```
I am not sure it's a good idea to use o_cc2 on a main container (especially with the missing o_colored_level) class... but maybe. In any case, I think there should be some uniformization accross the apps (cc @stefanorigano) to handle this. In events, it's white with blue cards, in slides the body color is darkened, here we now use o_cc2, ... and I am sure all the other apps have done it differently too.

I'll leave like this for now (just adding the o_colored_level class (which I'd like to get rid of at some points if possible...)).

qsm-odoo commented on addons/website_hr_recruitment/views/website_hr_recruitment_templates.xml:
```diff
@@ -3,81 +3,126 @@
                             </div>
-                        </a>
+                            <div class="col-lg-12 text-center pt24">
+                                <h5 class="fw-light text-muted">There are currently no open job opportunities,</h5>
+                                <h5 class="fw-light text-muted pb24">but feel free to <span class="fw-bold">contact us</span> for a spontaneous application.</h5>
```
A sentence split over two h5... not only this is bad for a good html structure but this is awful for translations...
I guess you did not like the line-height of a h5 with a <br> inside... but that would be the right way to go :grimacing: I'll go with that but with an extra mb class on the <br> to have what you wanted. We should probably just review all headings line-height though... @stefanorigano 

I also change from h5 to p.h5, don't like those being title.

In the end, by chance:
```xml
<h5>a</h5>
<h5>b</h5>
```
is the pixel-exact-equivalent as
```xml
<h5>a<br class="mb-2"/>b</h5>
```
but the second is a valid structure and translation-friendly.

qsm-odoo commented on addons/website_hr_recruitment/views/website_hr_recruitment_templates.xml:
```diff
@@ -3,81 +3,126 @@
+</template>
+
+<!-- Job Index Description -->
+<template id="job_index_description" name="Job Index Description">
```
I'll add a commit which only extracts this in a dedicated template first. Again: please think of the reviewer.

qsm-odoo commented on addons/website_hr_recruitment/views/website_hr_recruitment_templates.xml:
```diff
@@ -549,55 +578,80 @@
+                        <div class="row pt24 pb24">
+                            <div class="col pt24 pb24 text-center">
+                                <h1 class="fw-bolder">Congratulations!</h1>
+                                <h5 class="fw-light">
+                                    Your application has been posted successfully,
+                                </h5>
+                                <h5 class="fw-light pb24">
+                                    We usually respond within 3 days...
+                                </h5>
```
Another one :grimacing:

qsm-odoo reviewed (changes requested):
Finally taking the time and mustering the courage to review this :grimacing: 
Adding a batch of comments:
- when there are questions, I don't mind an answer :)
- when there is none, I am not asking to change anything: I am doing it myself (please do not push-force this branch anymore). Should I die tonight (:laughing:) I pushed my WIP of the first two commits on master-website-hr-recruitment-imp-cvs-qsm for now.

I'll finish everything about this tomorrow / Thursday :+1:

bso-odoo commented on addons/website_hr_recruitment/views/website_hr_recruitment_templates.xml:
```diff
@@ -3,81 +3,126 @@
+    </t>
+    <div class="o_job_infos d-flex flex-column">
+        <span t-field="job.address_id" class="fw-light" t-options='{"widget": "contact", "fields": ["city"], "no_tag_br": True, "null_text": "Remote"}'/>
+        <t t-if="str(job.department_id.name) != 'False'">
```
Nothing, I should have spotted those `str(...) != 'False'`.

qsm-odoo commented on addons/website_hr_recruitment/views/website_hr_recruitment_templates.xml:
```diff
@@ -549,55 +578,80 @@
+                                <img class="img-fluid o_we_custom_image" t-attf-style="width:{{ '50%' if (job and job.user_id) else '25%' }} !important;" src="/website_hr_recruitment/static/src/img/job_congratulations.svg" alt="Congratulations!" />
+                                <div class="row" id="o_recruitment_thank_cta">
+                                    <div class="col-lg-12 text-center mt32 mb48">
+                                        <h5 class="fw-light">In the meantime,</h5>
+                                        <h3 class="mt8 mb32 fw-bold">Take a look around our website:</h3>
```
h5 + h3 :grimacing:

qsm-odoo reviewed (changes requested):
Another batch of comments about second commit, also pushed my adaptation on master-website-hr-recruitment-imp-cvs-qsm.

Hopefully, tomorrow, I'll have the time to finish the remaining commits and merge

qsm-odoo commented on addons/website_hr_recruitment/controllers/main.py:
```diff
@@ -30,17 +30,32 @@ def sitemap_jobs(env, rule, qs):
-        domain = request.website.website_domain()
-        job_ids = Jobs.search(domain, order="is_published desc, sequence, no_of_recruitment desc").ids
-        # Browse jobs as superuser, because address is restricted
-        jobs = Jobs.sudo().browse(job_ids)
```
Removing jobs but still used below... but not if I take the second commit... ok let's squash.

qsm-odoo commented on addons/website/tests/test_fuzzy.py:
```diff
@@ -225,7 +225,7 @@ def test_06_case_insensitive_results(self):
     def test_07_no_fuzzy_for_mostly_number(self):
         """ Ensures exact match is used when search contains mostly numbers. """
-        self._create_page('Product P7935432254U7 page', 'Product P7935432254U7', '/numberpage')
+        self._create_page('Product P7935432254U7 page', 'Product P7935432254U7 kangaroo shoes', '/numberpage')
```
Moving this commit as first commit otherwise the PR is broken for a few commits if I understand the commit message right

qsm-odoo reviewed (changes requested):
Push-forcing after this batch of comments and continuing to review from my version of the commits from now on. Also stopping speaking to myself :shrug:

qsm-odoo commented on addons/website_hr_recruitment/controllers/main.py:
```diff
@@ -3,73 +3,188 @@
+
+        # Deduce offices, departments and countries offices
+        def sort(items, key):
+            """ Sort items alphabetically followed by None if present
```
I'll make that sentence clearer, maybe "Sort items alphabetically followed by **a** None **value** if present **in the given list**"

qsm-odoo commented:
@xmo-odoo 

> Why is a second description field necessary

A second description is necessary as those description are not used as the same place and for the same thing:
- The existing one holds the content of a job description page
- The new one holds a small description displayed on the /jobs page

> and why does it need non-whitelisted attributes, to say nothing of embedded forms and form fields?

I am not sure of what you mean... are you asking why we need to use `sanitize_attributes=False` and `sanitize_form=False`? The reason for that is like for all the other website HTML fields: we want those areas to be as editable as possible by the users who have the edition rights (and considering these are not website visitors). This keeps the "base" sanitization. Although, this should actually be marked with `sanitize_overridable=True` as the other description field, missed that in the rebase.

qsm-odoo commented:
Hello @xmo-odoo, any news? :wink:

rdeodoo commented:
The reason is that some of our snippets are using form/input.
The base sanitize would prevent those snippets to be used in those fields as those would be striped, which we don't want.
But at the same time, we still want to sanitize the field for the other elements.
As @qsm-odoo said: `we want those areas to be as editable as possible by the users who have the edition rights` but also keep security as high as possible.
(By extension, it's also why we want to add the new `sanitize_override` property: let admin do everything he wants but not restricted users.)

Task: https://www.odoo.com/web#id=2209554&cids=1&model=project.task&view_type=form
PR: https://github.com/odoo-dev/odoo/commit/388c222c6c4bb7e2fe3e67009b248359ae0fd3db

rdeodoo commented:
> OK but do we have a use for these in here?
> [..]
> What is the use case for putting per-job forms on /jobs?

It's an open question, I'd personally not drag & drop any snippet containing an input inside that part, but why not?
User might like to drop there a newsletter block, or a search bar snippet, those are the 2 coming in my mind without checking but you got the idea: those are totally legit and valid flows, despite being weird.
It might not be what most people do, but some will, it would be weird to have those blocks not working in this particular field, because we judged it was not the best place to use those snippets.
There is probably even some more legit snippets including a `<input/>` to place in this new shot description field.

All in all, I don't think we should add the `sanitize_form=False` property case by case: if the field is a "website builder" editable field (meaning he can drag & drop stuff inside), give it the form/input capability.

> whatever garbage the snippets contain

There is no garbage involved here, it's just legit `<input/>`: search bar snippet, newsletter snippet, donation snippet, form snippet, discussion subscribe snippet etc etc

------------

At the end, your question is purely functional: is there a functional reasonable use case which would led the user to need a `<input/>` element in the new short description of the job?
I'd say:
1. Yes and no: clearly not the main use case but some people will want to do it
2. There is no reason to not do it, keep it simple and add the form/input capability to every field being editable in the website builder. It will just bring confusion to the end user if there is some places where he can, and some where he can't.
3. We might later come up with another snippet using a `<input/>` element which would have a totally legit use case in this field (might even be the case in some community snippets?)

What's your opinion?

----------

> If the goal is to allow whatever garbage the snippets contain, why not call the flags that, at least it's clear what the goal is.

The goal is to remove any form/input, which is exactly what it does. It is implemented low level so anyone can use it should they need it. It simply call `lxml.clean` with `forms=False` at the end, the naming seems nice to me.
I would not include any hint about `snippets` or anything in its name.

rdeodoo commented:
@odoo/rd-security Hello, this one is waiting for your blessing.

mart-e commented:
> Hello, this one is waiting for your blessing.

Well I think @xmo-odoo has given a review (remove sanitize_form and add sanitize_overridable) and I tend to agree with him. Sooooo well, no blessing from me I am afraid...

tivisse commented:
@robodoo override=ci/security

Rationale: After a discussion with @mart-e and considering that the basic job description field is already ``sanitize_form=False``, this is ok from a consistency point of view. 

On the other hand, if it becomes automatic to authorize forms in html fields then we should consider authorizing it by default. Because the current API is not clear and it looks strange to sanitize field where we release 5 options each time.

PS : Don't forget to apply the 2 other minor security points.

Thanks for your work @karlosvalverde 

cc @qsm-odoo @rdeodoo

qsm-odoo reviewed (approved):
@robodoo rebase-ff r+

robodoo commented:
Merge method set to rebase and fast-forward.
